### DISCUSSION TITLE
Lesson 7: testUserCanFundInteractions test failed.

### DISCUSSION THREAD

**Discussion by 78carla** [Cyfrin/foundry-full-course-cu#405] [Q&A] [answered]
Hi, 
following the video step by step, Patric typed this piece of code in the InetractionTest.t.sol file

``` solidity
function testUserCanFundInteractions() public {
       
        FundFundMe fundFundMe = new FundFundMe();

        vm.prank(USER);
        vm.deal(USER, 1e18);

        fundFundMe.fundFundMe(address(fundMe));
        
        address funder = fundMe.getFunders(0);
        console.log("Funder: %s", funder);
        assertEq(funder, USER);
    }
```

and this in the Interaction.s.sol file

``` solidity
    function fundFundMe(address mostRecentlyDeployed) public{
        FundMe(payable (mostRecentlyDeployed)).fund{value: SEND_VALUE}();
        console.log("FundMe funded with %s", SEND_VALUE);
    }
```

The timestamp of the particular slide in the course video is around 2:08:01. 
    
I tried to run the test using:
```
forge test --match-test testUserCanFundInteractions -vvv
```
and I had this error:
```
[⠰] Compiling...
No files changed, compilation skipped

Running 1 test for test/interaction/InteractionTest.t.sol:InteractionsTest
[FAIL. Reason: EvmError: Revert] testUserCanFundInteractions() (gas: 298806)
Traces:
  [298806] InteractionsTest::testUserCanFundInteractions() 
    ├─ [248508] → new FundFundMe@0x2a9e8fa175F45b235efDdD97d2727741EF4Eee63
    │   └─ ← 1020 bytes of code
    ├─ [0] VM::prank(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D]) 
    │   └─ ← ()
    ├─ [0] VM::deal(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], 1000000000000000000 [1e18]) 
    │   └─ ← ()
    ├─ [9826] FundFundMe::fundFundMe(FundMe: [0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3]) 
    │   ├─ [0] FundMe::fund{value: 100000000000000000}() 
    │   │   └─ ← "EvmError: OutOfFund"
    │   └─ ← "EvmError: Revert"
    └─ ← "EvmError: Revert"

Test result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 4.11ms
Ran 1 test suites: 0 tests passed, 1 failed, 0 skipped (1 total tests)

Failing tests:
Encountered 1 failing test in test/interaction/InteractionTest.t.sol:InteractionsTest
[FAIL. Reason: EvmError: Revert] testUserCanFundInteractions() (gas: 298806)

Encountered a total of 1 failing tests, 0 tests succeeded
```

I tried finding a solution using AI and here on the other discussions but without success.
Anyone can point me in the right direction?

Thanks in advance,

**Comment by Naman1729** [comment]
Hey @78carla , 

In InteractionsTest.t.sol ,
```solidity
    function testUserCanFundInteractions() public {
        FundFundMe fundFundMe = new FundFundMe();

        fundFundMe.fundFundMe(address(fundMe));

        address funder = fundMe.getFunders(0);
        console.log("Funder: %s", funder);
        assertEq(funder, msg.sender);
    }
```
In Interaction.s.sol ,
```solidity
        function fundFundMe(address mostRecentlyDeployed) public {
             vm.startBroadcast();
             FundMe(payable(mostRecentlyDeployed)).fund{value: SEND_VALUE}();
             vm.stopBroadcast();
         }
```
put vm.Broadcast() in Interaction.s.sol and replace the user by msg.sender. then, it will work.

**Reply by 78carla** [reply]
Hi, 
thanks a lot for your reply. 
The code that you have provided to me is different from the one written by Patrick during lesson 7. 
My question is: how can I run the test using the same code as Patrick (copied and pasted by me on top of this thread) without having the "outOfFund" error? The test seems to pass with success for @PatrickAlphaC so I'm trying to resolve the issue and have the same result... :)

**Reply by shrutic11** [reply]
I had similar concern as well. But I rewatched the video multiple times and I noticed that Patrick started writing the code differently and ended it differently. It doesn't work if you follow it as it is in the video. There is some disconnect. Please follow the suggested change to get the test passing.

**Reply by Naman1729** [reply]
Hello @78carla,
Could you kindly provide the timestamp of the video to which you are referring. This will help us to locate and see this section in video.

**Reply by shrutic11** [reply]
Sure. Write your test as given between 2.08.35 and 2.09.09. Make sure your setUp() is also coded as Patrick has it.
And then, you should have the following in your Interactions.s.script

```solidity     
   function fundFundMe(address mostRecentlyDeployed) public {
             vm.startBroadcast();
             FundMe(payable(mostRecentlyDeployed)).fund{value: SEND_VALUE}();
             vm.stopBroadcast();
         }
````

**Reply by 78carla** [reply]
Hi @Naman1729,
I have written the timestamp of the video in my first message but I'm adding it here again for your reference. 
The timestamp of the particular slide in the course video is around 2:08:01.
Thanks,

**Reply by 78carla** [reply]
Hi @shrutic11,
I re-watched the video several times too before asking here :). From my understanding, there are 2 different tests: 

1.  the first test is about "fund". Patrick run the test, he had an error and he adjusted the test using the _prank_ and _deal_ cheatcode. I'm referring to this test and it ends at 2:08:01. I assumed that this test should work because he had gone ahead with the second test.
2. the second test is about the "withdraw" and he has modified also the "fund" part. This works fine for me. It ends at 2:09:02

**Reply by ChocLabEdi** [reply]
Hi @Naman1729 ,

While your suggestion works  "Interaction.s.sol and replace the user by msg.sender ", could you elaborate who's msg.sender in this case? The results of test seems to suggest 'DefaultSender'.  

Also, if we were to test with a specified USER, any suggestion on how to write the test? I have a Javascript background and am successful with the below tweaks to Patrick's code to make the test work, but seems less than ideal to have to create a second parameter in function fundFundMe() just to accommodate the test. What's your thought?

In Interactions.s.sol:
```
 function fundFundMe(address mostRecentlyDeployed, address testUSER) public {
        if (testUSER != address(0)) {
            vm.prank(testUSER);
            FundMe(payable(mostRecentlyDeployed)).fund{value: SEND_VALUE}();
            console.log('Logging testing route to make sure USER in InteractionsTest.t.sol is indeed passed in here');
        } else {
        FundMe(payable(mostRecentlyDeployed)).fund{value: SEND_VALUE}();
        }
        console.log("Funded FundMe contract with %s", SEND_VALUE);
    }
```

In InteractionsTest.t.sol:
```
address public USER = makeAddr("user");
uint256 constant SEND_VALUE = 0.1 ether;
uint256 constant STARTING_BALANCE = 10 ether; 

function setUp() external {
        fundMe = new DeployFundMe().run();
        vm.deal(USER, STARTING_BALANCE);
    } 

 function testUserCanFundInteractions() public {
        FundFundMe fundFundMe = new FundFundMe();
        fundFundMe.fundFundMe(address(fundMe), USER);

        address funder = fundMe.getFunder(0); //index 0
        assertEq(funder, USER);
}
```

**Reply by Abbasjafri-syed** [reply]
you can use ```testUSER``` like this ```address public USER = makeAddr("user");```

**Comment by Esmio** [comment]
function testUserCanFundInteractions() public {
        FundFundMe fundFundMe = new FundFundMe();
        fundFundMe.fundFundMe(address(fundMe));

        address funder = fundMe.getFunder(0);
        assertEq(funder, msg.sender);

        WithdrawFundMe withdrawFundMe = new WithdrawFundMe();
        withdrawFundMe.withdrawFundMe(address(fundMe));

        assert(address(fundMe).balance == 0);
    }
I think the this test goes fund first and then withdraw

**Comment by XAudit174176** [comment]
Hi everyone, I've finished lesson 7, and I think beginners like me who are trying to write and deploy things like what Patrick did, will face some issues ... exactly on the function: `testUserCanFundInteractions` 
Please watch the video again with more attention, and you see exactly before running tests, video jumps somehow and the codes are a little changed
So I believe it could be a practice for you to move yourself a little forward and write this test by your own
Nothing is 100% perfect, but if you followed this up till now, you have the passion to move forward and not stop at this point, right?

BTW, if you wanted to ignore this or you couldn't fix the script and test for Interactions tests, these are mine: (nothing new, but it works)

<h3>Interactions.s.sol</h3>

```
// SPDX-Licence-Identifier: MIT

pragma solidity ^0.8.18;

import {Script, console} from "forge-std/Script.sol";
import {DevOpsTools} from "foundry-devops/src/DevOpsTools.sol";
import {FundMe} from "../src/FundMe.sol";

contract FundFundMe is Script {
    uint256 constant SEND_VALUE = 0.01 ether;

    function fundFundMe(address mostRecentlyDeployed) public {
        vm.startBroadcast();
        FundMe(payable(mostRecentlyDeployed)).fund{value: SEND_VALUE}();
        vm.stopBroadcast();
        console.log("Funded FundMe contract with %s", SEND_VALUE);
    }

    function run() external {
        address mostRecentlyDeployed = DevOpsTools.get_most_recent_deployment(
            "FundMe",
            block.chainid
        );
        fundFundMe(mostRecentlyDeployed);
    }
}

contract WithdrawFundMe is Script {
    
    function withdrawFundMe(address mostRecentlyDeployed) public {
        vm.startBroadcast();
        FundMe(payable(mostRecentlyDeployed)).withdraw();
        vm.stopBroadcast();
    }

    function run() external {
        address mostRecentlyDeployed = DevOpsTools.get_most_recent_deployment(
            "FundMe",
            block.chainid
        );
        withdrawFundMe(mostRecentlyDeployed);
    }
}
```

<h3>Interactions.t.sol</h3>

```
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.18;

import {Test, console} from "forge-std/Test.sol";
import {FundMe} from "../../src/FundMe.sol";
import {DeployFundMe} from "../../script/DeployFundMe.s.sol";
import {FundFundMe, WithdrawFundMe} from "../../script/Interactions.s.sol";

contract InteractionsTest is Test {

    FundMe fundMe;

    address USER = makeAddr("user");

    uint256 constant SEND_VALUE = 0.1 ether;
    uint256 constant STARTING_BALANCE = 10 ether;
    uint256 constant GAS_PRICE = 1;

    function setUp() external {
        DeployFundMe deploy = new DeployFundMe();
        fundMe = deploy.run();
        vm.deal(USER, STARTING_BALANCE);
    }

    function testUserCanFundInteractions() public {
        FundFundMe fundFundMe = new FundFundMe();
        fundFundMe.fundFundMe(address(fundMe));

        WithdrawFundMe withdrawFundMe = new WithdrawFundMe();
        withdrawFundMe.withdrawFundMe(address(fundMe));

        assert(address(fundMe).balance == 0);
        
    }
}
```

Let's go to the next step...

**Reply by 00decree** [reply]
expanding on this

I think patrick edited two different version of video together in this part of the lesson, hence the confusion.

the problem in the video, the code shown for `fundFundMe()` and `withdrawFundMe()` in `Interactions.s.sol` is missing start/stop broadcast. (he moved it to `run()` in the video)

and since we are calling `fundFundMe()` and  `withdrawFundMe()` from `InteractionsTest.t.sol` not the `run()` method, we need to add the start/stop broadcast to the `fundFundMe()` and  `withdrawFundMe()` method in `Interactions.s.sol` , so that the tx is broadcasted.

**Reply by dt6120** [reply]
@00decree I am unable to understand how adding `vm.broadcast()` solved the `OutOfFund` error. If possible, can you please give some more info on it? Thanks!

**Reply by uddercover** [reply, marked as answer]
@dt6120 Also for anyone else confused. Remember earlier in the course when we learnt about the vm keyword, Patrick mentioned that ```vm.startBroadcast()``` changes the sender of a function back to the original ```msg.sender``` (yourself) as opposed to the contract calling the function.
This means the person calling fund is you, the original sender and as foundry gives a lot of test ether to your fake address, you have lots of funds. Hence, the ```OutOfFund``` error disappears. :>

**Reply by Hamobcdev** [reply]
thank you guys, got really mixed up here and went through the video 5 or 6 times, it does seem that there are multiple edits, still not sure what the issue was, after entering https://github.com/XAudit174176 code that enabled the pass but will go through again to see where I went wrong. Thanks. Hmm those mixed up broadcasts?? Ah well moving along now, thank you all.

**Reply by EngrPips** [reply]
> @dt6120 Also for anyone else confused. Remember earlier in the course when we learnt about the vm keyword, Patrick mentioned that `vm.startBroadcast()` changes the sender of a function back to the original `msg.sender` (yourself) as opposed to the contract calling the function. This means the person calling fund is you, the original sender and as foundry gives a lot of test ether to your fake address, you have lots of funds. Hence, the `OutOfFund` error disappears. :>

This was marked as answer @adesokan247

**Comment by dt6120** [comment]
I also had a doubt related to this test case: how are we able to send `SEND_VALUE` to fundFundMe even though the function is non-payable? I am assuming the `OutOfFund` error is because even though the call to `fundFundMe` was passed no value, we are still trying to send `SEND_VALUE` to `fund`.

**Reply by Abbasjafri-syed** [reply]
because the function 'fund' is payable in the contract fundme..

**Comment by Abbasjafri-syed** [comment]
> Hi, following the video step by step, Patric typed this piece of code in the InetractionTest.t.sol file
> 
> ```solidity
> function testUserCanFundInteractions() public {
>        
>         FundFundMe fundFundMe = new FundFundMe();
> 
>         vm.prank(USER);
>         vm.deal(USER, 1e18);
> 
>         fundFundMe.fundFundMe(address(fundMe));
>         
>         address funder = fundMe.getFunders(0);
>         console.log("Funder: %s", funder);
>         assertEq(funder, USER);
>     }
> ```
> 
> and this in the Interaction.s.sol file
> 
> ```solidity
>     function fundFundMe(address mostRecentlyDeployed) public{
>         FundMe(payable (mostRecentlyDeployed)).fund{value: SEND_VALUE}();
>         console.log("FundMe funded with %s", SEND_VALUE);
>     }
> ```
> 
> The timestamp of the particular slide in the course video is around 2:08:01.
> 
> I tried to run the test using:
> 
> ```
> forge test --match-test testUserCanFundInteractions -vvv
> ```
> 
> and I had this error:
> 
> ```
> [⠰] Compiling...
> No files changed, compilation skipped
> 
> Running 1 test for test/interaction/InteractionTest.t.sol:InteractionsTest
> [FAIL. Reason: EvmError: Revert] testUserCanFundInteractions() (gas: 298806)
> Traces:
>   [298806] InteractionsTest::testUserCanFundInteractions() 
>     ├─ [248508] → new FundFundMe@0x2a9e8fa175F45b235efDdD97d2727741EF4Eee63
>     │   └─ ← 1020 bytes of code
>     ├─ [0] VM::prank(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D]) 
>     │   └─ ← ()
>     ├─ [0] VM::deal(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], 1000000000000000000 [1e18]) 
>     │   └─ ← ()
>     ├─ [9826] FundFundMe::fundFundMe(FundMe: [0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3]) 
>     │   ├─ [0] FundMe::fund{value: 100000000000000000}() 
>     │   │   └─ ← "EvmError: OutOfFund"
>     │   └─ ← "EvmError: Revert"
>     └─ ← "EvmError: Revert"
> 
> Test result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 4.11ms
> Ran 1 test suites: 0 tests passed, 1 failed, 0 skipped (1 total tests)
> 
> Failing tests:
> Encountered 1 failing test in test/interaction/InteractionTest.t.sol:InteractionsTest
> [FAIL. Reason: EvmError: Revert] testUserCanFundInteractions() (gas: 298806)
> 
> Encountered a total of 1 failing tests, 0 tests succeeded
> ```
> 
> I tried finding a solution using AI and here on the other discussions but without success. Anyone can point me in the right direction?
> 
> Thanks in advance,

you can use testnet rpc ... ```forge test --mt testUserCanFundInteractions -vvv --f $testnet_rpc``` it works for me

**Comment by zachcomeau** [comment]
Hi! I'm brand new to coding and happy I was able to figure this out own my own. I was just stuck in the SAME EXACT spot. 

I was able to get this test to pass by adding the vm.startbroadcast and vm.stopbroadcast further upstream in the fundFundMe function in Interactions.s.sol.

just thinking about how this script worked I thought it would make sense to fund the address earlier. 

here is an example of my code 

```solidity 
contract FundFundMe is Script {
    uint256 constant SEND_VALUE = 0.01 ether;
    
    function fundFundMe (address mostRecentlyDeployed) public {
       vm.startBroadcast(); 
       FundMe(payable(mostRecentlyDeployed)).fund{value: SEND_VALUE}();
       vm.stopBroadcast();
       console.log("Funded FundMe %s", SEND_VALUE);

```

**Reply by bhanupbalusu** [reply]
I think this issue 'EvmError: OutOfFund' still persists. Even I used vm.startBroadcast() as mentioned in above comment but it did not resolve the issue. I got the following error. 

├─ [0] VM::startBroadcast()
    │   │   └─ ← you have an active prank; broadcasting and pranks are not compatible
    │   └─ ← you have an active prank; broadcasting and pranks are not compatible
    └─ ← you have an active prank; broadcasting and pranks are not compatible

Test result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 1.61ms

Has anyone found resolution? Thanks.

**Reply by bhanupbalusu** [reply]
Ok I removed vm.prank and it is working now. Thanks @Naman1729 for the solution.

But I have a query related to this, when I executed in debug mode I am seeing that 'msg.sender' is 'InteractionsTest' and having a balance of '7.922e28'. When any contract (in this case - 'Interactions.Test' as msg.sender) is deployed it's balance should be 0 right? Why is it having initial balance of '7.922e28'?

Here is the logs - 
├─ [0] console::log(InteractionsTest: [0x34A1D3fff3958843C43aD80F30b94c510645C316]) [staticcall]
 │   │   └─ ← (
│   ├─ [0] console::log(79228162514264337593543950335 [7.922e28]) [staticcall]

Any explanation is appreciated. Thank you.

**Comment by saintjer2224** [comment]
@78carla I got this OutOfFund issue in FundFundMe sorted out by removing these two lines below from the InteractionsTest.t.sol:
        //address funder = fundMe.getFunder(0);
        //assertEq(funder, USER); //test is only checking if funder is same as user 
        
I've also done the broadcast on the fundFundMe function and not on the run function as originally showed by Patrick.
        
My understanding is we are testing if user is able to fund FundMe. When it does, then script prints "Funded FundMe with xxxx". It really is able to do that with the following loc:

***
INTERACTIONSTEST.T.SOL

```
//SPDX-License-Identifier: MIT

pragma solidity ^0.8.19;

import {Test, console} from "forge-std/Test.sol";
import {FundMe} from "../../src/FundMe.sol";
import {DeployFundMe} from "../../script/DeployFundMe.s.sol";
import {FundFundMe} from "../../script/Interactions.s.sol"; 

contract InteractionsTest is Test {
    FundMe fundMe;

    address USER = makeAddr("user");
    uint256 constant SEND_VALUE = 0.5 ether; 
    uint256 constant STARTING_BALANCE = 10 ether; 
    uint256 constant GAS_PRICE = 1;

    function setUp() external {
        DeployFundMe deploy = new DeployFundMe();
        fundMe = deploy.run(); 
        vm.deal(USER, STARTING_BALANCE);
    }

    function testUserCanFundInteractions() public {
        FundFundMe fundFundMe = new FundFundMe();
        fundFundMe.fundFundMe(address(fundMe));

    }
}

```

***
INTERACTIONS.S.SOL

```
//SPDX-License-Identifier: MIT

pragma solidity ^0.8.19;

import {Script, console} from "forge-std/Script.sol";
import {DevOpsTools} from "lib/foundry-devops/src/DevOpsTools.sol";
import {FundMe} from "../src/FundMe.sol";

contract FundFundMe is Script {
    uint256 constant SEND_VALUE = 0.1 ether;

    function fundFundMe(address mostRecentlyDeployed) public {
        console.log(
            "Account balance before funding: %s",
            address(this).balance
        );
        vm.startBroadcast();
        FundMe(payable(mostRecentlyDeployed)).fund{value: SEND_VALUE}();
        vm.stopBroadcast();
        console.log("Funded FundMe with %s", SEND_VALUE);
    }

    function run() external {
        address mostRecentlyDeployed = DevOpsTools.get_most_recent_deployment(
            "FundMe", 
            block.chainid 
        );

        fundFundMe(mostRecentlyDeployed);

    }
}
}

````

OUTPUT

```
Logs:
  Account balance before funding: 0
  Funded FundMe with 100000000000000000
  
Test result: ok. 1 passed; 0 failed; 0 skipped; finished in 2.47ms
```

**Comment by adesokan247** [comment]
@78carla  have you been able to resolve it
